sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp

```diff
@@ -31,6 +31,8 @@ namespace Azure { namespace Storage { namespace Blobs {
    */
   struct TagAccessConditions
   {
+    virtual ~TagAccessConditions() = default;
```

Comment:
What's the motivation of change like this? I have difficulty understanding how this will benefit developers or customers or anybody.

Reply:
@Jinming-Hu, if you derive from a class without a virtual destructor, you introduce a memory leak. It is not a matter of guideline, it is a bug.

See for yourself (https://onlinegdb.com/HyxceW4OO):
```
#include <iostream>
#include <memory>

class CString
{
public:
    CString() { std::cout << "CString constructed, allocating 10 KB\n"; }
    ~CString() { std::cout << "CString destructed, releasing 10 KB\n"; }
};

class BaseClass
{
public:
    //virtual ~BaseClass() = default;
};

class Derived : public BaseClass
{
public:
    CString str;
};

int main()
{
    std::unique_ptr<BaseClass> smartPtr(new Derived());
    return 0;
}
```

With `virtual ~BaseClass() = default;` commented:
```
CString constructed, allocating 10 KB

...Program finished with exit code 0
```

With `virtual ~BaseClass() = default;` uncommented:
```
CString constructed, allocating 10 KB
CString destructed, releasing 10 KB

...Program finished with exit code 0   
```

Reply:
If you say, inherited classes never stored as `Base*`, then the constructor can be made `protected non-virtual`: then everything will compile as long as your statement is true. And at the moment someone tries to delete the `Base*` the problematic code that introduces a memory leak won't compile.

Reply:
@antkmsft I understand the scenario in your example. The problem is that virtual destructor is useful only when you delete a derived class via based pointer. But this doesn't always happen in C++. 


> It is not a matter of guideline, it is a bug.

In the case of `TagAccessConditions`, `BlobAccessConditions` inherits from `TagAccessConditions`. It's a bug when people delete `BlobAccessConditions` via a pointer to `TagAccessConditions`. But in storage code, nobody is supposed to allocate `BlobAccessConditions` on heap, and nobody is supposed to assign this object to `TagAccessConditions` pointer.


What you're trying to do looks to me like, it's a bug if we forget to `free` some memory allocated by `malloc`. But `malloc` itself is not a bug. Forgetting to free it is a bug. So to avoid this kind of bug, we completely forbid `malloc`.

Reply:
I understand, but it can be customer code that does that. In any case, if you think public virtual destructor in that case is a problem, then there should be no problem if we declare it as protected non-virtual to prevent memory leak from ever happening.

Reply:
> I understand, but it can be customer code that does that. In any case, if you think public virtual destructor in that case is a problem, then there should be no problem if we declare it as protected non-virtual to prevent memory leak from ever happening.

My concern is not about public or protected. My concern is that we're taking precautions against things that are never going to happen.

Reply:
@CaseyCarter Please also comment whether it's a common practice in C++ world to take this kind of precautionary measures against something that are never supposed to happen.